### PR TITLE
fix: #MZI-136 fix display names

### PR DESCRIPTION
**Pull request by Matiyeu**
## Describe your changes

Displaying usernames correctly when sending mail from Zimbra expert.
Remove duplicates of the same user.

## Checklist tests

- [x] List inbox/outbox with external and internal sender/receiver

## Issue ticket number and link

[JIRA#MZI-136](https://jira.support-ent.fr/browse/MZI-136)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 1]
```diff
@@ -141,6 +142,52 @@ private void processListMessages(JsonObject zimbraResponse,
         processSearchResult(zimbraMessages, frontMessages, addressMap, result);
     }
 
+    private Future<Void> addDisplayNames(JsonArray frontMsg) {
+        Promise<Void> promise = Promise.promise();
+
+        JsonArray userIds = new JsonArray();
+        for (Object obj : frontMsg) {
+            JsonObject mail = (JsonObject) obj;
+            JsonArray displayNames = mail.getJsonArray(FrontConstants.MAIL_DISPLAYNAMES);
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: mettre un new JsonArray par défaut

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 2]
```diff
@@ -141,6 +142,52 @@ private void processListMessages(JsonObject zimbraResponse,
...
+            for (Object d : displayNames) {
+                String userId = ((JsonArray) d).getString(0);
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: Vérifier la taille de d pour éviter les index out of bound

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 3]
```diff
@@ -141,6 +142,52 @@ private void processListMessages(JsonObject zimbraResponse,
...
+                Map<String, String> usernames = new HashMap<>();
+                for (JsonObject user : users) {
+                    String id = user.getString("id");
+                    String username = user.getString("username");
+                    usernames.put(id, username);
+                }
+
+                for (Object obj : frontMsg) {
+                    JsonObject mail = (JsonObject) obj;
+                    JsonArray displayNames = mail.getJsonArray(FrontConstants.MAIL_DISPLAYNAMES);
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: idem

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 4]
```diff
@@ -141,6 +142,52 @@ private void processListMessages(JsonObject zimbraResponse,
...
+                    for (Object d : displayNames) {
+                        String userId = ((JsonArray) d).getString(0);
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: idem

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 5]
```diff
@@ -259,6 +307,15 @@ private void processMessageMultipart(JsonObject msgFront, JsonArray multiparts)
         msgFront.put(MESSAGE_ATTACHMENTS, mparts.getAttachmentsJson());
     }
 
+    private boolean displayNamesContainsUserId(JsonArray jsonArray, String userId) {
+        for (Object object : jsonArray) {
+            JsonArray array = (JsonArray) object;
```
Comment (sjollois) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java:343]: risque d'index out of bound

**Review by sjollois — APPROVED** [review]

**Review by sjollois — APPROVED** [review]

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 6]
```diff
@@ -141,6 +142,54 @@ private void processListMessages(JsonObject zimbraResponse,
...
+
+        userService.getUsers(userIds, userIds, (response) -> {
+            if (response.isLeft()) {
+                promise.fail(response.left().getValue());
+            } else {
+                List<JsonObject> users = response.right().getValue().getList();
+
+                Map<String, String> usernames = new HashMap<>();
+                for (JsonObject user : users) {
+                    String id = user.getString("id");
```
Comment (aymericdevriesere) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: magic string

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 7]
```diff
@@ -141,6 +142,54 @@ private void processListMessages(JsonObject zimbraResponse,
...
+                    String username = user.getString("username");
```
Comment (aymericdevriesere) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: magic string

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java [hunk 8]
```diff
@@ -141,6 +142,54 @@ private void processListMessages(JsonObject zimbraResponse,
...
+            for (Object displayName : displayNames) {
+                if (!(displayName instanceof JsonArray) || ((JsonArray) displayName).size() == 0) continue;
+                String userId = ((JsonArray) displayName).getString(0);
+                if (userId != null && StringHelper.isUUID(userId)) {
+                    userIds.add(userId);
+                }
+            }
+        }
+
+        userService.getUsers(userIds, userIds, (response) -> {
```
Comment (aymericdevriesere) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java:200]: externalisable ?

File: zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java (hunk 6 above)
Comment (aymericdevriesere) [inline zimbra/src/main/java/fr/openent/zimbra/service/impl/MessageService.java]: verif qu'ils soit non nul

**Review by aymericdevriesere — APPROVED** [review]

**Pull request merged**